### PR TITLE
Feature #2701 ismn

### LINKED ISSUE #2701: Enhance ASCII2NC to read ISMN point observations of soil moisture and temperature
## Describe the New Feature ##
This issue is to enhance the MET ASCII2NC tool to ingest point observations of soil moisture and temperature.

Sample data can be retrieved from https://ismn.bafg.de/en/data/data-download.

As described on the [ISMN formats page](https://ismn.bafg.de/en/data/formats/), data is available in two formats:
1. Following the [CEOP conventions](https://ismn.bafg.de/en/data/ceop-standard/) (Coordinated Energy and water cycle Observations Project).
2. Stored in a [Header+values](https://ismn.bafg.de/en/data/header-value-files/) format.

Adding support for either of these is acceptable. @anewman89, please advise on which is preferable.

Note that in both cases, the variable name is embedded only in the filename:
```
CSE_Network_Station_Variablename_depthfrom_depthto_sensorname _startdate_enddate.ext
```
The **Variablename** element defines the data type and separate variables are provided in separate files.

Be sure that ASCII2NC can handle multiple input files in each run.

### Acceptance Testing ###
Add at least one new ascii2nc unit test to demonstrate this functionality.
Coordinate on the corresponding METplus use case.

### Time Estimate ###
3 days (?)

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
No sub-issues needed.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
7790111 (CLASP)

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [x] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- Once created, link to the corresponding METplus Use Case issue.
- For that METplus Use Case issue, recommend updating the METplus [Verification Datasets Guide](https://metplus.readthedocs.io/en/latest/Verification_Datasets/index.html) to describe ISMN data.

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
Adds new `-format ismn` option for ASCII2NC.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
- Pulled lots of ISMN data from https://ismn.bafg.de/en/data/.
- Manually ran through ascii2nc to confirm that its processed as expected.
- Manually ran the newly added test for ascii2nc for processing SNOTEL data:
```
/Volumes/d1/projects/MET/MET_development/MET-feature_2701_ismn/internal/test_unit/../../share/met/../../bin/ascii2nc \
    -format ismn  \
    /Volumes/d1/projects/MET/MET_unit_test/MET_test_input/obs_data/ismn/SNOTEL/*/*.stm \
    /Volumes/d1/projects/MET/MET_unit_test/MET_test_output/ascii2nc/ismn_SNOTEL_20220924_2022027.nc
```

- Plotted the result:

```
../../bin/plot_point_obs /Volumes/d1/projects/MET/MET_unit_test/MET_test_output/ascii2nc/ismn_SNOTEL_20220924_2022027.nc ismn_SNOTEL.ps
```
- Confirmed SNOTEL dots are in the expected spots:
Screen Shot 2023-12-06 at 3 17 51 PM

- Used Python utility to dump obs to ascii:

```
python3 share/met/python/utility/print_pointnc2ascii.py /Volumes/d1/projects/MET/MET_unit_test/MET_test_output/ascii2nc/ismn_SNOTEL_20220924_2022027.nc > ismn_SNOTEL.txt
```

- Inspected some of the results in 
[ismn_SNOTEL.txt.gz](https://github.com/dtcenter/MET/files/13592537/ismn_SNOTEL.txt.gz):

```
SNOTEL AGUA_CANYON          20220924_000000 37.5200 -112.2700 2712.7200 SNOD  NA  0.0000             C01  -0.02540000
SNOTEL AGUA_CANYON          20220924_010000 37.5200 -112.2700 2712.7200 SNOD  NA  0.0000             C01  -0.02540000
SNOTEL AGUA_CANYON          20220924_020000 37.5200 -112.2700 2712.7200 SNOD  NA  0.0000               G   0.00000000
SNOTEL AGUA_CANYON          20220924_030000 37.5200 -112.2700 2712.7200 SNOD  NA  0.0000               G   0.00000000
SNOTEL AGUA_CANYON          20220924_040000 37.5200 -112.2700 2712.7200 SNOD  NA  0.0000               G   0.00000000
```
- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
- Review code changes.
- Review documentation changes, optionally via [RTD for this PR](https://met--2758.org.readthedocs.build/en/2758/Users_Guide/reformat_point.html).
- Consider whether levels should be reported as positive or negative? Currently, below ground is positive. For example, soil moisture:

`SNOTEL INDEPENDENCE_CAMP    20220924_050000 39.4500 -120.2900 2134.5100 SOILW NA  0.0508               G   0.09500000`

Above ground is negative. For example, 2-m air temperature:

`SNOTEL WARD_MOUNTAIN        20220924_230000 39.1300 -114.9600 2804.1599 TMP   NA -2.0000               G 290.54998779`

Above, the `-2` means 2-meter temperature. That's a potential source of confusion. Should the signs be swapped so that above ground is positive and below ground is negative?

- Could run manually yourself on seneca using:
```
/d1/projects/MET/MET_pull_requests/met-12.0.0/beta3/MET-feature_2701_ismn/internal/test_unit/../../share/met/../../bin/ascii2nc \
    -format ismn  \
    /d1/projects/MET/MET_test_data/unit_test/obs_data/ismn/SNOTEL/*/*.stm \
    /d1/projects/MET/MET_pull_requests/met-12.0.0/beta3/MET-feature_2701_ismn/internal/test_unit/../../test_output/ascii2nc/ismn_SNOTEL_20220924_2022027.nc
```
- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [x] Do these changes include sufficient testing updates? **[Yes]**

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
- Adds 1 new unit test output file `ascii2nc/ismn_SNOTEL_20220924_2022027.nc`
- [x] Please complete this pull request review by **[Monday 12/11/23]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by JohnHalleyGotway** [comment]
As expected, the only difference in the [GHA automated regression test](https://github.com/dtcenter/MET/actions/runs/7121420804) for this pull request is the addition of 1 new output file:
```
dir1: /data/output/met_test_truth contains 1126 files
dir2: /data/output/met_test_output contains 1127 files

ERROR: folder /data/output/met_test_truth missing 1 files
    ascii2nc/ismn_SNOTEL_20220924_2022027.nc 
```

**Comment by JohnHalleyGotway** [comment]
@anewman89, FYI, the operating system on seneca was updated last Friday, and I fixed the `development.seneca` environment file to handle those changes yesterday. Since this PR is still open, I just recompiled this feature branch on seneca in `/d1/projects/MET/MET_pull_requests/met-12.0.0/beta3/MET-feature_2701_ismn` in the updated environment.

Would you like to meet sometime this week to discuss the changes and step through the process of doing the PR review?

**Comment by anewman89** [comment]
@JohnHalleyGotway Thanks for the recompile.  Yes, it might be good to meet, I'll email you.

File: docs/Users_Guide/reformat_point.rst [hunk 1]
```diff
@@ -539,7 +541,7 @@ Required Arguments for ascii2nc
 Optional Arguments for ascii2nc
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-3. The **-format ASCII_format** option may be set to "met_point", "little_r", "surfrad", "wwsis", "airnowhourlyaqobs", "airnowhourly", "airnowdaily_v2", "ndbc_standard", "aeronet", "aeronetv2", "aeronetv3", or "python". If passing in ISIS data, use the "surfrad" format flag.
+3. The **-format ASCII_format** option may be set to "met_point", "little_r", "surfrad", "wwsis", "airnowhourlyaqobs", "airnowhourly", "airnowdaily_v2", "ndbc_standard", "ismn", "aeronet", "aeronetv2", "aeronetv3", or "python". If passing in ISIS data, use the "surfrad" format flag.
```
Comment (JohnHalleyGotway) [inline docs/Users_Guide/reformat_point.rst:544]:
```suggestion
3. The **-format ASCII_format** option may be set to "met_point", "little_r", "surfrad", "wwsis", "airnowhourlyaqobs", "airnowhourly", "airnowdaily_v2", "ndbc_standard", "ismn", here's an edit "aeronet", "aeronetv2", "aeronetv3", or "python". If passing in ISIS data, use the "surfrad" format flag.
```

**Comment by JohnHalleyGotway** [comment]
@anewman89 I'd like to get these changes merged prior to the holidays and to prevent this feature branch from drifting too far away from the develop branch.

Are you available to do the review this week or should I reassign to someone else?

**Comment by JohnHalleyGotway** [comment]
@anewman89 and @DanielAdriaansen, please advise whether you can complete this PR review by Fri, Jan 19th, 2024 for inclusion in the MET-12.0.0-beta3 release. If not, I'll reassign to another reviewer. Thanks.

**Review by anewman89 — APPROVED** [review]
@JohnHalleyGotway These changes look good to me.  Thanks and sorry for the delay!

**Pull request merged**